### DISCUSSION TITLE
Discussion for Documentation

### DISCUSSION THREAD

**Discussion by sehilyi** [gosling-lang/gosling.js#24] [General]
The thread of this discussion will cover the questions and comments around the geminid grammar to make documentation.

**Comment by wangqianwen0418** [comment]
The current documentations [undergoing 📝 ] is available at  https://github.com/sehilyi/geminid/wiki/Docs

Some comments after trying the basic:
- 👍 overall, the structure is clear, should be easy for those who have prior experience in declarative visualisation grammar. Not sure about the domain experts, who maybe unfamiliar with visualization grammars
- ❓ the part about `track.data` and `track.metadata` is a little bit confusing. Not sure what are the supported data formats 
- ❓ how can I rotate a chart, for example, the area chart in basic marks, 90 degree? I have tried to switch the encoding of x and y axis but failed. maybe this is a rare case in genomic visualization?)
- ❓ it is a little bit confusing for me to understand the difference between `rect` and `bar`. Also confused about the encoding of width and height of these two marks
- ❓ I didn't see the legend (when set legend: true) of color when `"color": {"type": "quantitative", "legend": true}`
- 💬 one track has different options based on the value of  `track.mark`, need good documentations to explain what are the unique options of each mark and what are the shared options
- 💬 it makes more sense to me if we move `width`, `height`, and `static` to the `layout` option
- 💬 As a user, I may want to add a title and some necessary descriptions in the visualization, not just in the code. Maybe it will help if `description` is an object with options such as `title`, `subtitle`, `tooltip`.

**Comment by sehilyi** [comment]
- Entire grammar: https://github.com/sehilyi/geminid/blob/master/src/core/geminid.schema.ts
- How each mark is rendered (encoded): https://github.com/sehilyi/geminid/tree/master/src/core/mark

**Comment by sehilyi** [comment]
For each subsection (e.g., "Point"), we can cover the following contents:
- Short description
- A table (just like the current `layout` table) (if applicable)
- Example
   - Part of spec (if applicable)
   - Actual example embedded (if applicable)

**Comment by wangqianwen0418** [comment]
Some questions:
1. is `size` option also supported when the mark is not `point`?
2. `strokeWidth` is currently listed below `encoding`, is it just a style option, or it can be bound with a data field? The same question goes to `opacity`. 
3. I am little confused about `x`, `xe`, and `x1e`. Would you mind providing more explanations?

**Reply by sehilyi** [reply]
Thats good question.
1. `size` is being used for other marks as well (e.g., when `size` is provided, the width of a `bar` is encoded with the size, which is defined [here](https://github.com/sehilyi/geminid/blob/c273e5640e64766d84e7744c5c178f02986eb88f/src/core/mark/bar.ts#L196)).
2. My intension was to support encoding quantitative data with `strokeWidth`, but did not find a good use case for that. So, currently it does not encode data ([this](https://github.com/sehilyi/geminid/blob/c273e5640e64766d84e7744c5c178f02986eb88f/src/core/geminid-track-model.ts#L288) is the part of source code that determines whether to encode data or not, FYI).
3. For most of the cases, `x` w/ an additional channel, `xe`, are used ("e" refers to "end"). However, band connection (`link`) should consider four different channels (start and end positions of two genomic regions), and this is the only case where `x1` and `x1e` are used.

**Reply by wangqianwen0418** [reply]
Thanks !
1. but `size` may not work for `area` and `line`? Would you list the marks where size encoding is supported?
2. I see the type of `opacity` is ChannelValue, so it doesn't support data encoding?
3. does `axis` only works for `x`, `y`, `xe`, `ye`?
4. would you mind listing the visual channels that support `legend`?

**Reply by sehilyi** [reply]
1. `point`, `rect`, `bar`, `triangle-*`.
2. That is correct. (I think we can decide now whether to move some of the channels to the `style` option). What do you think of moving `opacity`, `background`, and even `strokeWidth` to the `style` option?
3. `axis` is used in `x` and `y`
4. Currently, only `row` and `color` whey they are encoded with `nominal` values. Will need to add more `legend`.

**Reply by wangqianwen0418** [reply]
Thanks for the reply!
I think we can safely move `background` and `strokeWidth` to the style option. Note sure about `opacity`, I feel sometimes I may want to encode the opacity of points. 
But we can move `opacity` to `style` for the current version. No need to support all functions at the first version

**Reply by sehilyi** [reply]
Good suggestion. In that case, I think we can move only `background` to the `style`, and support encoding quantitative values to `opacity` and `strokeWidth`, which only requires me to add a few lines. I think it does not hurt to have that. I've added an issue for this (https://github.com/sehilyi/geminid/issues/39).

**Comment by wangqianwen0418** [comment]
Another question:
is `compositeMark` still supported? or you plan to replace it with `superpose`?

**Reply by sehilyi** [reply]
Surprised that you found this :)

Although I called it `compositeMark`, it is more like a reusable template for custom tracks. It is not being used in the grammar currently (once supported but removed temporally to focus on implementing other features), but I am thinking of adding an option in `Track` (like `template`) and moving the functionality of `compositeMark` to this option.

**Comment by wangqianwen0418** [comment]
Hi @sehilyi, I am trying to make a table (like the table below) in the wiki docs.


mark type | supported visual channels
-- | --
point | x, y, size, color, strokeWidth, opacity
line | x, y, color, strokeWidth, opacity
rect | x, xe, color, strokeWidth, opacity
bar | x, y, ...
area | x, y, ...
...|...

Since you are more familiar with the code base, would you mind helping me fulfill the skeleton of the table (if you think this table is useful)?

**Reply by wangqianwen0418** [reply]
Also, I am planning to make two tables about the visual channel properties: 

---

**properties shared by all visual channels**
visual channel properties | type | description
-- | -- | --
field | string | the data field name
type | string | specify type of the data field. support "genomic", "nominal", "quantitative"
aggregate | string | support "max", "min", "mean", "bin", "count"
domain |   |  
range |   |  

**properties that are only used by certain visual channels**
visual channel properties | type | supported visual channels| description
-- | -- | -- | --
axis | string |  `x`, `y` | It specifies the position of the axis. Support "none", "top", "bottom", "left", "right"
baseline
legend

**Reply by sehilyi** [reply]
They look great! I will update the \<supported visual channels\> table directly in the wiki.

**Reply by wangqianwen0418** [reply]
Many thanks for updating the supported visual channel table~

Just one question, it seems `size` is also supported in the line mark (and some other marks that don't include `size` in the table, I haven't check all of them). [here](<https://sehilyi.github.io/geminid/?full=false&spec=(9'NtleJBasic%20Marks%3A%20Line'4'subNtleJTutorial%20ExWs'4'layoutJlinear'4'arrangement~direcNonJverNcal'G'rowSizes6180G'columnSizes6800)4'tracksj9*(9CdataXurlJhttps%3A%2F%2Fresgen.io%2Fapi%2Fv1%2FNleset_info%2F%3Fd%3DUvVPeLHuRDiYA3qwFlm7xQIKNleset'OmetadataXKhiglass-mulNvecIrowJsWIcolumnZvalueJpeakIcategoriesj'sW%201'%5DOmarkJline'4CxXfieldZKgenomicIdomain~chromosomeJ1'G'intervalj1G3000500%5D)4*CaxisJbottom'Oy~fieldJpeak'G'KquanNtaNve')4Csize~value62)9*)9%5D%0A)*%20%204%2C96!%209%0A*C**'G%2C%20I'4*CJ6'KtypeJNtiO9**)4CWampleX6(9*CZJposiNonIj6%5B~6('%01~jZXWONKJIGC964*_>)

**Comment by wangqianwen0418** [comment]
I am not very sure about the `rule` mark, could you give more explanation?

**Reply by sehilyi** [reply]
Currently, it is only used for showing segmented lines (w/ diverse styles) that represent genomic intervals (like in gene annotation tracks). But, I am considering to either (1) merge its encoding functionalities to `rect` since they are similar each other or (2) add more unique encodings to `rule`. (Originally, it was for showing vertical or horizontal lines.)

Since work is left for this mark type, how about leaving this part empty in the docs for now and revisit this later after implementation.

**Reply by wangqianwen0418** [reply]
Great, I agree. I will leave this part empty.

**Comment by wangqianwen0418** [comment]
In the current grammar, users define the layout of a track using ```"circularLayout": true/false```. Maybe ```"trackLayout": "circular"/"linear"``` makes more sense?

**Reply by sehilyi** [reply]
I agree, and I was thinking of renaming some properties for the consistency, including `circularLayout` and `static`. I've just added a related issue (#59).

**Comment by wangqianwen0418** [comment]
One more thing, what do you think about adding a render button in the online editor. 
The right side only updates after the button is clicked

The current online editor renders whenever there is a change in the left side. Sometimes, the whole webpage crashes due to some rendering error while I am modifying the configuration. That is a little bit annoying.

**Reply by sehilyi** [reply]
Hi Qianwen, thanks for the suggestion. I was thinking about adding a mode-change button ("manual" <=> "auto") for the use case.

Also, can you share me what kinds of changes on the code resulted in the app crash? It looks like a bug, it should show the following message rather than the app crash:

![Screen Shot 2020-12-28 at 2 35 48 PM](https://user-images.githubusercontent.com/9922882/103238876-06621280-491a-11eb-875c-10f37184e8f0.png)

FYI, I've fixed a bug that resulted in the app crash three days ago (https://github.com/sehilyi/geminid/issues/52). If you confronted the issue since then, please let me know.

**Reply by wangqianwen0418** [reply]
Oh, it is exactly the bug you fixed at #52~ Thanks!

**Comment by sehilyi** [comment]
Hi @wangqianwen0418, FYI, you can now generate unique urls (#51) to display custom specs in the editor. This can be useful when providing links for individual examples (i.e., "Open the example in online editor" in the current docs). To obtain the url, edit the spec as you want and click on the button on the top-right corner in the editor. Then, a unique url will be copied to your clipboard.

The example url looks like this:

[Open the example in online editor][1]

[1]: https://sehilyi.github.io/geminid/?full=false&spec=(%0AZlayoutK(V'jlinear'%2CV'directionK'vertical'%2CV'columnSizesK800%2CV'rowSizesK60%0A*)%2C%0AZtracksK%5BV(VZdataK(J'urlK'https%3A%2F%2Fraw.githubusercontent.com%2Fsehilyi%2Fgemini-datasets%2Fmaster%2Fdata%2FUCSC.HG38.Human.CytoBandIdeogram.csv'Q'jcsv'Q'c%3DFieldK'C%3D'Q'%40FieldsK%5B'%60LchromEnd'%5DV*%2BuperposeK%5BJ(JZmarkK'text'QRMtrue)~textK(XNameLjnominal')QZcolorK(J%26QY%22'%3FL%3FL%3FL%3FLwhiteL%3F'~visibilityK(J*ZoperationK'less-than'Q*ZconditionK('widthK'%7Cxe-x%7CLtransitionPaddingK10)Q*ZtargetK'mark'J*)QZstyleK('textS%7BK0%25rect'QRMtrue)~colorK(J%26QY%22J**Zwhite%3BD9D9D9%3B979797%3B636363'Q**Z%3F%3BA0A0F2'J**%5DJ*%5E-r'QRJ***M%24'q%3E~%3C%5E-l'QRJ***M%24'p%3E~%3C)J)V*%5D%2CVZxK(JX%60'Q'j%40'Q'domainK('c%3DK'1')Q'axisK'top'V*)%2CVZxeK(XchromEndLj%40'%2Bize_20%2Btroke_'gray'%2B%7B_0.5%2BtyleK('outlineK'white')V)%0A*%5D%0A)*%20%20JV**K!%20L'%2C%20'M(XStainLoneOfK%5B'acen'%5D%2C%20'notKQ%2CJRZdataTransformK(J*ZfilterK%5BV%0A**X'fieldK'Y*ZdomainK%5B'gnegLgpos25Lgpos50Lgpos75Z*'_K('valueKjtypeK'~%5DJ*)QZ%22Lgpos100Lgvar'%5DQ*ZrangeK%5B%24false)Q***(XNameLincludeK%25)J)Q(JZmarkK'%26**XStain'Q*Zjnominal'%2B)%2CVZs%3B'Q**Z%23%3Ccolor_'%23B40101'%3Dhromosome%3ELnotKfalse)J**%3Fblack%40genomic%5E%25triangle%60chromStart%7BtrokeWidth%01%7B%60%5E%40%3F%3E%3D%3C%3B%2B%26%25%24%22~j_ZYXVRQMLKJ*_

**Comment by wangqianwen0418** [comment]
Some question about `bar`: 
how the width of a bar is calculated?  
can user set the gap between bars? 
does `x` indicates the left side of a bar or the centre of a bar?

many thanks~

**Reply by sehilyi** [reply]
`x` indicates the center position of bars.

- You could use both `x` and `xe` to bind start and end position of bars to a certain genomic interval.
- You could also use `size` to make the width of bars being constant values (e.g., `size: { value: 6}` == width of bars is 6px).

**Comment by wangqianwen0418** [comment]
This seems to be a small bug. The annotation of the y axis is wired.
<img width="344" alt="image" src="https://user-images.githubusercontent.com/19774198/104112331-9198bc00-52bb-11eb-9042-063ef5fda1da.png">


Should I start an issue?

**Reply by sehilyi** [reply]
Yes, it would be really helpful if you write issues whenever you find unexpected behaviors

**Reply by sehilyi** [reply]
I guess that axis appeared when you set `axis: "left"`? It would be great if you could put a spec that made this behavior

**Comment by wangqianwen0418** [comment]
I am a little confused about `triangle-d`. Is it actually `triangle-b`, i.e., triangle bottom?

**Reply by sehilyi** [reply]
It refers to "down", but `b` for bottom might be a more consistent term

**Reply by sehilyi** [reply]
@wangqianwen0418 Do you think it is better to use "triangle-up", "triangle-down", "triangle-right", or "triangle-left" as in Vega? (Rather than the short names, like "triangle-l").

**Comment by wangqianwen0418** [comment]
@sehilyi , when you are convenient, would you mind taking a look at the current wiki documentations? You comments are welcome!

I haven't start writing the section `data` and the table `properties that are only used by certain visual channels` , so you can just ignore those parts.

It would be great if you can point me to some examples of the `semantic zoom` and `tooltip`. 

Do you think we have some examples of using `triangle` marks without introducing `superpose`? I try to introduce the types of mark before `superpose` but find it is hard to achieve

**Reply by sehilyi** [reply]
Hi @wangqianwen0418, thank you for support, I will look into the documentation.

I think the current best example for `triangle` w/o using `superpose` would be showing gene position, like <a href="https://sehilyi.github.io/geminid/?full=false&spec=(V'layoutB'linear'%2CV'ar~gementB('directionB'vertical'GcolumnSizesB800GrowSizesB110)%2CV'tracksRV*(9'dataj'urlB'https%3A%2F%2Fhiglass.io%2Fapi%2Fv1%2Ftileset_info%2F%3Fd%3DOHJakQICQD6gTD7skx4EWA'K'Ytileset'9Zmetadataj'Yhiglass-bed'K'genomicX1%242%22K'valueX5Mst~d'GN)KL3Mname'GN)9*%5DK'exonIntervalX12%2413%229)K*'dataT~sformj**'filterR9****(Utype'GoneOfR'geneqK***(Ust~d'GoneOfR'%2Bq9***%5D9**)K*'markB'triangle-r'K*'xj**Uend'K**'Ygenomic'K**'domainB('chromosomeB'1')K**'axisB'top'9**)K*'sizeB('valueB20ZcolorjUst~d'K'NK'domainR'%2B'G-'%5DK'~geR'%237585FF'G%23FF8A85'%5D9Zvisibilityj'operationB'less-than'K'conditionB('widthB'%7Cxe-x%7C'Gt~sitionPaddingB10)K'targetB'mark'9ZopacityB('valueB0.5)V*)V%5D%0A)*%20%209V**B!%20G%2C%20'K%2C9*L*('indexBMGnameB'NYnominal'RB%5BU'fieldB'V%0A*XFieldsR9*LYtypeB'Z)%2C9'jB(9*q'%5DGnotBfalse)~ran%22Mend')9*%5D%24Mstart')KL%01%24%22~qjZYXVURNMLKGB9*_">this one</a>. Even better one would be showing a 90-deg rotated heatmap (https://github.com/sehilyi/geminid/issues/33), which is not yet supported in the grammar.

Regarding `semantic zoom`, the upper three track definitions in the "Semantic Zooming" demo will be good such examples. If you have question regarding the examples, please let me know.

The `tooltip` example is <a href="https://sehilyi.github.io/geminid/?full=false&spec=(G'layoutblinear'%2CG'arrangementJ('directionbverticalBrowSizesJ180%2C6columnSizesJ800)%2CG'tracksjG*(N'tooltipjN*(6MstartBXWStartCendBXWEndC%22WEndCTWPeak'%20)N%5D9'dataJ(N*'urlbhttps%3A%2F%2Fresgen.io%2Fapi%2Fv1%2Ftileset_info%2F%3Fd%3DUvVPeLHuRDiYA3qwFlm7xQOItileset'N)9'metadataJ(N*'Ihiglass-multivecOrowbKOcolumnbpZOvaluebpeakOcategoriesj'~1B~2B~3B~4'%5DN)9'markbrect'9'xJ(N*'MpZOXOdomainJ('chromosomeb1Bintervalj1%2C%203000500%5D)9*'axisbtop'N)9'rowJ('M%22BlegendJtrue)9'colorJ('MT')G*)G%5D%0A)*%20%206%20'9%2CNB'%2C6C%20PZ'%20)9*(6MG%0A*ItypebJ!%20KsampleMfieldbNG**O'9*'TpeakBIquantitativeWBaltbXIgenomicZositionb!6jJ%5B~K%20%22KBInominal%01%22~jbZXWTONMKJIGCB96*_">here</a> (Please find the `tooltip: {...}` definition in the spec. And please note that, currently, the functionality of `tooltip` is quite limited as only `rect` marks in `linear` layouts are supported for showing tooltips.)

**Reply by wangqianwen0418** [reply]
Many thanks!

**Reply by wangqianwen0418** [reply]
Since the current support for `tooltip` is limited, do you think we should remove it from the current documentation?

**Comment by sehilyi** [comment]
@wangqianwen0418 I've briefly looked into the documentation and made some changes, but it looks good! I can see that people can already understand many parts of the grammar by reading the docs.

Thinks we may need to consider adding in the future would be (1) describing visual parameters for circular layouts (e.g., `startAngle`, `endAngle`, `innerRadius`, `outerRadius`, ...) and (2) little more details about field/channel (e.g., `field.type`, `channel.domain`, `channel.range`).

Also, as documentation covers many parts already, it would be worth discussing about static sites at some point.

**Comment by wangqianwen0418** [comment]
Do you think the triggerCondition interface can be refactored in the following way:

```javascript
interface TriggerCondition {
  operation: LogicalOperation,
  measure: "width"|"height"|"zoomLevel",
  threshold: number|"|xe-x|",
  target: 'track' | 'mark' | 'superpose',
  conditionPadding?:number,
  transitionPadding?:number
}
```

this interface looks more clear to me.  
But I totally understand that you might have other design considerations

**Reply by wangqianwen0418** [reply]
Also, I can create a pull request for this modification (later this week) if it makes sense to you

**Reply by sehilyi** [reply]
I like the renaming "condition" => "measure" & "threshold" which I think is more understandable. I also like the idea to force users to select a single `measure`.

One thing we could additionally do would be separating the interface based on the `measure` since its values (`width`, `height`, `zoomLevel`) use different properties. So something like:

```js
...
visibility?: VisibilityCondition;
...

type VisibilityCondition = SizeVisibilityConsition | ZoomLevelVisibilityCondition;

interface CommonVisibilityCondition {
  operation: LogicalOperation;
  target: 'track' | 'mark'
}

interface SizeVisibilityCondition extends CommonVisibilityCondition {
  measure: 'width' | 'height';
  threshold: number | '|xe-x|;
  conditionPadding?: number;
  transitionPadding?: number;
}

interface ZoomLevelVisibilityCondition extends CommonVisibilityCondition {
  measure: 'zoomLevel';
  threshold: number;
}
```

We could add an issue for this and update the code in a PR. It would be helpful if you can help on this.

**Reply by wangqianwen0418** [reply]
Cool, I have created an issue https://github.com/gosling-lang/gosling.js/issues/124~

**Comment by wangqianwen0418** [comment]
Just to clarify, the current dataTransform only supports three types of filters: `oneOf`, `inRange`, `include`, right?

Another question, do you think it makes sense to assign `not` a default false value in the filter?

**Reply by sehilyi** [reply]
That is correct, and I think making the `not` an optional value (`not?: boolean`) and considering it as `false` by default if it is `undefined` looks like a good idea.

**Reply by wangqianwen0418** [reply]
Cool~ I created a pull request.
I have also updated the wiki documents about `semantic zoom`, `dataTransform`, and add some details about the circular layout

**Reply by sehilyi** [reply]
Just read the documentation, and those parts are great!

**Comment by wangqianwen0418** [comment]
Just to clarify, does this mean that csv data can only be loaded from url while json data can only be loaded from inplace json value?
If a user has a csv file on his/her computer, how to access the csv file in Gosling?
```javascript
export interface CSVDataGosling extends DataDeepGoslingCommon {
    type: 'csv';
    url?: string;
    separator?: string;
    // experimental
    headerNames?: string[];
    chromosomePrefix?: string;
    longToWideId?: string;
}

export interface JSONDataGosling extends DataDeepGoslingCommon {
    type: 'json';
    values?: Datum[];
}
```

Also, could you point me to some examples using JSON data?
As Nils mentioned, I think it is important that we have data that is more accessible and easy to read, so that users can know how to play with their own data.

**Reply by wangqianwen0418** [reply]
Will we keep or remove metaData? I remember you mentioned before but not sure.

**Reply by sehilyi** [reply]
> Will we keep or remove metaData? I remember you mentioned before but not sure.

I think we will be using `metadata` at least until our version 1 release. This is a required option especially for converting complex genomic data formats into tabular formats.

**Reply by sehilyi** [reply]
> Also, could you point me to some examples using JSON data?

Yes, I will share a cytoband example with you on Slack, and try finding more examples. I think such examples will be important for users who just started to learn our grammar.

Just to give you a better sense on using the `json` data in the grammar, you can pick any example specs with a `csv` type data and change it to use `json` type data: change the `type` from `csv` to `json` in the data config, convert the csv file into a json array, and put the array to the `values` property (e.g., `values: [ {'chr': 'chr1', 'p1': 1, 'p2': 100000}]`).

**Reply by sehilyi** [reply]
> Just to clarify, does this mean that csv data can only be loaded from url while json data can only be loaded from inplace json value?

That is correct.

**Reply by sehilyi** [reply]
> If a user has a csv file on his/her computer, how to access the csv file in Gosling?

As a JSON-format grammar itself, we cannot support that. Since our grammar uses a JSON format as an input, raw data can be either (1) specified in a JSON format as a part of the spec or (2) stored at somewhere while its url is provided in the spec (I think this is the same in Vega/Vega-Lite).

When we build a wrapper library in the future (e.g., Gosling-React or Gosling.py), we could support uploading a local csv file. However, I think our main focus would be on the grammar itself until the paper submission.

Perhaps, an easiest way to allow people to play with their csv files would be supporting it in the editor (so outside of the grammar), e.g., a user upload a csv file, the editor converts it into a json format, and put it into a spec using a `json` type data spec.

One thing worth mentioning here is that our focus is more on the genomic-specific file formats than those generic formats, and, because such genomic data is commonly large, allowing people to upload the local file directly to a system is quite challenging. (HiGlass currently only support specifying urls of files for this reason, and no local files supported). We will need to think about more on how to allow people to load their local data for the more accessible and easier grammar usage, which I think is a really important issue.

> As Nils mentioned, I think it is important that we have data that is more accessible and easy to read, so that users can know how to play with their own data.

Yes, I totally agree. I think we can prepare datasets (both in AWS buckets and Gosling's public HiGlass servers) and provide a list of datasets with public urls and detailed information so that people can try our grammar with them more accessibly. We can also add tutorials using small csv/json files to more easily give a sense on how the data can be used in the grammar for visual encoding.

**Reply by sehilyi** [reply]
@wangqianwen0418 I have a question that is perhaps related to this thread:

Do you think it will be largely helpful for users to have a Data Table panel in the editor where people can check how the raw data looks? I think this could help on understanding what is the role of `metadata` option as well.

**Reply by wangqianwen0418** [reply]
Hi Sehi, so many thanks for these information!

> I think we will be using metadata at least until our version 1 release. This is a required option especially for converting complex genomic data formats into tabular formats.

The timeline looks good.

For the JSON & CSV data format, I would not worry too much about showing them in the editor (they are not bad in terms of readability). Users can always download the csv file from the url and open it on their own computer. But I do agree that examples using csv data are important for the tutorial.   
If you do not have enough time, maybe we can prioritise csv examples over json. 
Inserting json data makes the code very long. And json doesn't much more readability than csv file.

For table panel & file upload features, yes, they are quite important.   
On the other hand, I think examples with csv files would be great enough for the twitter release.
CSV examples enable people to understand the data and play with the example.
Meanwhile, metaData examples show that Gosling has the ability to process more complicated genomic-specific file

**Reply by sehilyi** [reply]
> On the other hand, I think examples with csv files would be great enough for the twitter release.

We can discuss further tomorrow, but I think it would be still important to prepare enough number/diversity of examples with genomic-specific files since we want to give impression that the grammar can support complex and large genomic data.
